docs/PLUGIN.md

```diff
@@ -0,0 +1,314 @@
+# Custom Plugins
+
+Custom plugins are a powerful way to extend GPUd's monitoring capabilities. They allow you to create custom health checks and monitoring scripts that integrate seamlessly with GPUd's component system.
+
+## Overview
+
+Plugins in GPUd serve two main purposes:
+
+1. **Initialization (init)**: One-time setup tasks that run when GPUd starts
+2. **Monitoring (component)**: Regular health checks that run periodically
+
+Plugins can be configured to run as:
+- Single components (one health check)
+- Multiple components (multiple health checks from a single configuration)
+- Initialization tasks (one-time setup)
+
+## Plugin Types
+
+### Init Plugins
+Init plugins are used for one-time setup tasks when GPUd starts. They're perfect for:
+- Setting up system configurations
+- Installing required dependencies
+- Initializing monitoring tools
+- Setting up logging
+
+### Component Plugins
+Component plugins run periodically to monitor system health. They can be:
+- **Single Component**: One health check with one configuration
+- **Multi-Component**: Multiple health checks from a single configuration, either specified directly or loaded from a file
+
+## Configuration Format
+
+Plugins are configured using YAML files. Here's the basic structure:
+
+```yaml
+plugin_name: string  # Required, max 128 chars
+type: string        # Required, one of: "init", "component", "component_list"
+run_mode: string    # Optional, defaults to "manual"
+timeout: duration   # Optional, defaults to 1 minute
+interval: duration  # Optional, must be >= 1 minute if specified
+
+# For component_list type
+component_list: string[]  # Required for component_list type, unless component_list_file is specified
+component_list_file: string  # Optional, path to file containing component list
+
+health_state_plugin:
+  steps:
+    - name: string  # Required
+      run_bash_script:
+        content_type: string  # Required, e.g. "plaintext", "base64"
+        script: string       # Required, bash script content
+```
+
+## Component List Format
+
+Each component in the list can be specified in one of four formats:
+
+1. **Full Format**: `name#run_mode:param`
+   - `name`: Component name
+   - `run_mode`: Optional run mode (auto, manual, once)
```

Comment:
+1 for `once`. Can we document `once` mode has not been implemented?

Something like:

```suggestion
   - `run_mode`: Optional run mode (possible values: auto, manual) -- "once" mode to be supported in the future
```